File: 25-adv-ds2/6-monotonic-stack-applications/main.java
// Monotonic Stack - Java example (Next Greater Element)

import java.util.*;

public class Main {
    public static int[] nextGreaterElements(int[] nums) {
        int[] result = new int[nums.length];
        Arrays.fill(result, -1);
        Deque<Integer> stack = new ArrayDeque<>();

        for (int i = 0; i < nums.length; i++) {
            while (!stack.isEmpty() && nums[stack.peek()] < nums[i]) {
                result[stack.pop()] = nums[i];
            }
            stack.push(i);
        }
        return result;
    }

    public static void main(String[] args) {
        int[] nums = {2, 1, 2, 4, 3};
        int[] res = nextGreaterElements(nums);
        System.out.println(Arrays.toString(res));  // Output: [4, 2, 4, -1, -1]
    }
}

File: 25-adv-ds2/6-monotonic-stack-applications/main.cpp
// Monotonic Stack - C++ example (Next Greater Element)

#include <iostream>
#include <vector>
#include <stack>

using namespace std;

vector<int> nextGreaterElements(const vector<int>& nums) {
    vector<int> result(nums.size(), -1);
    stack<int> s;

    for (int i = 0; i < nums.size(); ++i) {
        while (!s.empty() && nums[s.top()] < nums[i]) {
            result[s.top()] = nums[i];
            s.pop();
        }
        s.push(i);
    }
    return result;
}

int main() {
    vector<int> nums = {2, 1, 2, 4, 3};
    vector<int> res = nextGreaterElements(nums);
    for (int val : res)
        cout << val << " ";
    cout << endl;  // Output: 4 2 4 -1 -1
    return 0;
}
